src/Bicep.Core/SemanticModel/NameBindingVisitor.cs

```diff
@@ -110,30 +131,61 @@ private Symbol ValidateFunctionFlags(Symbol symbol, TextSpan span)
+
+            Symbol? foundSymbol;
+            if (string.IsNullOrEmpty(@namespace))
             {
-                // we found the symbol in the local namespace
-                return ValidateFunctionFlags(localSymbol, span);
+                // attempt to find name in the imported namespaces
+                var namespaceSymbol = FindNamespace(name);
+
+                if (namespaceSymbol != null)
+                {
+                    // namespace symbol found
+                    return ValidateFunctionFlags(namespaceSymbol, span);
+                }
+
+                // declarations must not have a namespace value, namespaces are used to fully qualify a function access.
+                // There might be instances where a variable declaration for example uses the same name as one of the imported
+                // functions, in this case to differentiate a variable declaration vs a function access we check the namespace value,
+                // the former case must have an empty namespace value whereas the latter will have a namespace value.
+                if (this.declarations.TryGetValue(name, out var localSymbol))
+                {
+                    // we found the symbol in the local namespace
+                    return ValidateFunctionFlags(localSymbol, span);
+                }
+
+                // attempt to find function in all imported namespaces
+                var foundSymbols = this.namespaces
+                    .Select(ns => ns.TryGetFunctionSymbol(name))
+                    .Where(symbol => symbol != null)
+                    .ToList();
+
+                if (foundSymbols.Count() > 1)
+                {
+                    // ambiguous symbol
+                    return new ErrorSymbol(DiagnosticBuilder.ForPosition(span).AmbiguousSymbolReference(name, this.namespaces.Select(ns => ns.Name)));
+                }
+
+                foundSymbol = foundSymbols.FirstOrDefault();
             }
+            else
+            {
+                // attempt to find name in the imported namespaces
+                var namespaceSymbol = FindNamespace(@namespace!);
 
-            // symbol does not exist in the local namespace
-            // try it in the imported namespaces
-
-            // match in one of the namespaces
-            var foundSymbols = this.namespaces
-                .Select(ns => ns.TryGetFunctionSymbol(name))
-                .Where(symbol => symbol != null)
-                .ToList();
+                if (namespaceSymbol == null)
+                {
+                    return new ErrorSymbol(DiagnosticBuilder.ForPosition(span).SymbolicNameDoesNotExist(name));
+                }
 
-            if (foundSymbols.Count() > 1)
-            {
-                // ambiguous symbol
-                return new ErrorSymbol(DiagnosticBuilder.ForPosition(span).AmbiguousSymbolReference(name, this.namespaces.Select(ns => ns.Name)));
+                foundSymbol = namespaceSymbol.TryGetFunctionSymbol(name);
```

Comment:
If `TryGetFunctionSymbol` returns null, rather than letting the outer block report the error, why not return a more specific error here like "unable to find 'ABC' in namespace 'XYZ'" (since you have the name at this point)?

Reply:
Updated